packages/kotti-table/src/logic/order.js

```diff
@@ -0,0 +1,76 @@
+import pick from 'lodash/pick'
+import {
+	setColumnsArray,
+	getColumnsArray,
+	getColumnIndex,
+	getColumnRealIndex,
+} from './column'
+
+export const defaultState = {
+	orderedColumns: [],
+}
+
+export const mutations = {
+	orderBefore(store, fromColumn, toColumn) {
+		const { state } = store
+
+		const fromIndex = getColumnIndex(state, fromColumn)
+		state._columnsArray.splice(fromIndex, 1)
+
+		const toIndex = getColumnRealIndex(state, toColumn)
+		state._columnsArray.splice(toIndex, 0, fromColumn)
+
+		state._columnsArray.forEach((col, index) => {
+			col.order = index
+			col.index = index
+		})
+
+		store.emit('orderChange', getOrderedColumns(state))
+		store.commit('updateColumns')
+	},
+	setOrderedColumns(store, columns = store.state.orderedColumns) {
+		setColumnsArray(store.state, 'orderedColumns', ['prop', 'order'], columns)
+		store.commit('updateColumns', {
+			emitChange: false,
+			refreshColumnArray: true,
+		})
+	},
+}
+
+export const getters = {
+	getOrderedColumns(state) {
+		return getColumnsArray(state, 'orderedColumns')
+	},
+}
+
+export function resolveColumnsOrder({ _columns = {}, orderedColumns = [] }) {
+	for (let col of orderedColumns) {
+		_columns[col.prop] = col.order || _columns[col.prop]
+	}
+	return Object.values(_columns)
+		.filter(({ _deleted }) => !_deleted)
+		.map((col, index) => {
+			return {
+				orderAdvantage: 'order' in col ? 1 : -1,
+				order: col.order || col.index,
+				index: col.index || index,
+				col,
+			}
+		})
+		.sort((a, b) =>
+			a.order < b.order
+				? -1
+				: a.order > b.order
+				? 1
+				: a.orderAdvantage + a.index - (b.orderAdvantage + b.index),
+		)
+		.map(({ col }, order) => {
+			col.order = order
+			col.index = order
+			return col
```

Comment:
```suggestion
			return {...col, order, index: order }
```

And delete all lines above. Then compact the arrow function. This has the advantage that it treats its arguments as immutable, preventing weird bugs.

Reply:
this is an intentional mutation
columns are referenced in the table's observable state, if I create a new one I would need to set it with Vue.set

the map only exists to return the same column after the mutation is applied
I could have forEach if it returned the original array

Reply:
I hope we won’t run into weird bugs because of this in the future. 😟